routers/repo/compare.go

```diff
@@ -0,0 +1,313 @@
+		prInfo.MergeBase, headCommitID, setting.Git.MaxGitDiffLines,
+		setting.Git.MaxGitDiffLineCharacters, setting.Git.MaxGitDiffFiles)
+	if err != nil {
+		ctx.ServerError("GetDiffRange", err)
+		return false
+	}
+	ctx.Data["Diff"] = diff
+	ctx.Data["DiffNotAvailable"] = diff.NumFiles() == 0
+
+	headCommit, err := headGitRepo.GetCommit(headCommitID)
+	if err != nil {
+		ctx.ServerError("GetCommit", err)
+		return false
+	}
+
+	prInfo.Commits = models.ValidateCommitsWithEmails(prInfo.Commits)
+	prInfo.Commits = models.ParseCommitsWithSignature(prInfo.Commits)
+	prInfo.Commits = models.ParseCommitsWithStatus(prInfo.Commits, headRepo)
+	ctx.Data["Commits"] = prInfo.Commits
+	ctx.Data["CommitCount"] = prInfo.Commits.Len()
+	if ctx.Data["CommitCount"] == 0 {
+		ctx.Data["PageIsComparePull"] = false
+	}
+
+	if prInfo.Commits.Len() == 1 {
+		c := prInfo.Commits.Front().Value.(models.SignCommitWithStatuses)
+		title = strings.TrimSpace(c.UserCommit.Summary())
+
+		body := strings.Split(strings.TrimSpace(c.UserCommit.Message()), "\n")
+		if len(body) > 1 {
+			ctx.Data["content"] = strings.Join(body[1:], "\n")
+		}
+	} else {
+		title = headBranch
+	}
+
+	ctx.Data["title"] = title
+	ctx.Data["Username"] = headUser.Name
+	ctx.Data["Reponame"] = headRepo.Name
+	ctx.Data["IsImageFile"] = headCommit.IsImageFile
+
+	headTarget := path.Join(headUser.Name, repo.Name)
+	ctx.Data["SourcePath"] = setting.AppSubURL + "/" + path.Join(headTarget, "src", "commit", headCommitID)
+	ctx.Data["BeforeSourcePath"] = setting.AppSubURL + "/" + path.Join(headTarget, "src", "commit", prInfo.MergeBase)
+	ctx.Data["RawPath"] = setting.AppSubURL + "/" + path.Join(headTarget, "raw", "commit", headCommitID)
+	return false
+}
+
+// CompareDiff show different from one commit to another commit
+func CompareDiff(ctx *context.Context) {
+	headUser, headRepo, headGitRepo, prInfo, baseBranch, headBranch := ParseCompareInfo(ctx)
+	if ctx.Written() {
+		return
+	}
+
+	nothingToCompare := PrepareCompareDiff(ctx, headUser, headRepo, headGitRepo, prInfo, baseBranch, headBranch)
+	if ctx.Written() {
+		return
+	}
+
+	if ctx.Data["PageIsComparePull"] == true {
+		pr, err := models.GetUnmergedPullRequest(headRepo.ID, ctx.Repo.Repository.ID, headBranch, baseBranch)
+		if err != nil {
+			if !models.IsErrPullRequestNotExist(err) {
+				ctx.ServerError("GetUnmergedPullRequest", err)
+				return
+			}
+		} else {
+			ctx.Data["HasPullRequest"] = true
+			ctx.Data["PullRequest"] = pr
+			ctx.HTML(200, tplCompareDiff)
+			return
+		}
+
+		if !nothingToCompare {
+			// Setup information for new form.
+			RetrieveRepoMetas(ctx, ctx.Repo.Repository)
+			if ctx.Written() {
+				return
+			}
+		}
+
+		headBranches, err := headGitRepo.GetBranches()
+		if err != nil {
+			ctx.ServerError("GetBranches", err)
+			return
+		}
+		ctx.Data["HeadBranches"] = headBranches
+	}
+
+	beforeCommitID := ctx.Data["BeforeCommitID"].(string)
+	afterCommitID := ctx.Data["AfterCommitID"].(string)
```

Comment:
```suggestion
	afterCommitID := ctx.Data["AfterCommitID"].(string)
	if beforeCommit, _ := ctx.Repo.GitRepo.GetCommit(beforeCommitID); beforeCommit != nil {
		beforeCommitID = beforeCommit.ID.String()
	}
	if afterCommit, _ := ctx.Repo.GitRepo.GetCommit(afterCommitID); afterCommit != nil {
		afterCommitID = afterCommit.ID.String()
	}
```

Reply:
I would think something like this would work to use short hashes.

Reply:
So that won't work - needs to be done somewhere else.

Reply:
https://github.com/go-gitea/gitea/blob/c73353b8a2917dc352419ebcef7e505258372a6e/routers/repo/compare.go#L78-L79

needs to change - perhaps using an additional test.

Reply:
OK I've added a trick to this - @saitho take a look feel free to revert.

Reply:
I'll check that tomorrow. :) What about moving that check into `ctx.Repo.GitRepo.IsCommitExist`?

Reply:
That would make `ctx.Repo.GitRepo.IsCommitExist` true for every branch, tag or otherwise valid ref including short sha's. I don't know enough about the IsCommitExist to ensure that that would be correct so I left it alone.

If you do the work to check this and it doesn't change the semantics too much then that's fine.

Reply:
What I would say is that you should keep the head/base branch names here instead of changing them to before and after.

Reply:
Realized I did that in the wrong place. Oops.

Reply:
> That would make `ctx.Repo.GitRepo.IsCommitExist` true for every branch, tag or otherwise valid ref including short sha's.

Valid point. I'll leave it as it is right now and add a small comment that the line is for checking short hashs. ^^